src/lib/FreeC/Pass/KindInferencePass.hs

```diff
@@ -0,0 +1,211 @@
+checkConDecl (IR.ConDecl _ _ types) = mapM_ checkType types
+
+-- | Checks whether the type schema in a type signature is kind-correct
+checkTypeSig :: IR.TypeSig -> Converter ()
+checkTypeSig (IR.TypeSig _ _ typeSchema) = checkTypeSchema typeSchema
+
+-- | Checks whether a given type schema is kind-correct
+checkTypeSchema :: IR.TypeSchema -> Converter ()
+checkTypeSchema (IR.TypeSchema _ _ typ) = checkType typ
+
+-- | Checks whether a function declaration has kind-correct type annotations
+checkFuncDecl :: IR.FuncDecl -> Converter ()
+checkFuncDecl (IR.FuncDecl _ _ _ varPats retType rhs) = do
+  checkVarPats varPats
+  checkMaybeType retType
+  checkExpr rhs
+  return ()
+
+-- | Checks whether all type annotations in an expression are kind-correct
+checkExpr :: IR.Expr -> Converter ()
+checkExpr (IR.Con _ _ typeSchema      ) = checkMaybeTypeSchema typeSchema
+checkExpr (IR.Var _ _ typeSchema      ) = checkMaybeTypeSchema typeSchema
+checkExpr (IR.App _ lhs rhs typeSchema) = do
+  checkExpr lhs
+  checkExpr rhs
+  checkMaybeTypeSchema typeSchema
+checkExpr (IR.TypeAppExpr _ lhs rhs typeSchema) = do
+  checkExpr lhs
+  checkType rhs
+  checkMaybeTypeSchema typeSchema
+checkExpr (IR.If _ cond thenExp elseExp typeSchema) = do
+  checkExpr cond
+  checkExpr thenExp
+  checkExpr elseExp
+  checkMaybeTypeSchema typeSchema
+checkExpr (IR.Case _ scrutinee alts typeSchema) = do
+  checkExpr scrutinee
+  checkAlts alts
+  checkMaybeTypeSchema typeSchema
+checkExpr (IR.Undefined _ typeSchema      ) = checkMaybeTypeSchema typeSchema
+checkExpr (IR.ErrorExpr  _ _ typeSchema   ) = checkMaybeTypeSchema typeSchema
+checkExpr (IR.IntLiteral _ _ typeSchema   ) = checkMaybeTypeSchema typeSchema
+checkExpr (IR.Lambda _ args rhs typeSchema) = do
+  checkVarPats args
+  checkExpr rhs
+  checkMaybeTypeSchema typeSchema
+
+-- | Performs @checkAlt@ on all elements of the given list.
+checkAlts :: [IR.Alt] -> Converter ()
+checkAlts = mapM_ checkAlt
+
+-- | Checks whether the variable patterns have correct type annotations and the
+--   expression has kind-correct type annotations
+checkAlt :: IR.Alt -> Converter ()
+checkAlt (IR.Alt _ _ varPats rhs) = do
+  checkVarPats varPats
+  checkExpr rhs
+
+-- | Performs @checkVarPat@ on all elems in the given list.
+checkVarPats :: [IR.VarPat] -> Converter ()
+checkVarPats = mapM_ checkVarPat
+
+-- | Checks whether the type annotation of a variable pattern is kind-correct.
+checkVarPat :: IR.VarPat -> Converter ()
+checkVarPat (IR.VarPat _ _ typ) = checkMaybeType typ
+
+-- | Checks whether a type is kind-correct
+checkType :: IR.Type -> Converter ()
+checkType = checkCorrectArities
+
+-- | Checks if all type constructors have the correct number of arguments
+checkCorrectArities :: IR.Type -> Converter ()
+checkCorrectArities = checkCorrectArities' 0
+
+-- | Helper function for @checkCorrectArities@
+checkCorrectArities' :: Int -> IR.Type -> Converter ()
+checkCorrectArities' depth (IR.TypeVar srcSpan varId) =
```

Comment:
Please add a brief explanation to the comment what the purpose of the additional argument is.